setuptools/extension.py

```diff
@@ -126,6 +126,12 @@ class Extension(_Extension):
       specified on Windows. (since v63)
     """
 
+    # These 4 are set and used in setuptools/command/build_ext.py
+    _full_name = ""  #: Private API, internal use only.
+    _links_to_dynamic = False  #: Private API, internal use only.
+    _needs_stub = False  #: Private API, internal use only.
+    _file_name = ""  #: Private API, internal use only.
```

Comment:
There is one very edge-y case where it is useful to not define default values before hand: to avoid people forgetting to call `finalize_options` if they modify the extension list like in https://github.com/pypa/setuptools/issues/4529.

Tricky... In this case I feel like would like to have the cake (static type annotations) but eat it too (have the exception happening if someone forgets to call `finalize_options`).

Reply:
Maybe wrap in a `if TYPE_CHECKING`?

Reply:
One could also declare the attribute type without a default value:
```py
>>> class A:
...   foo: str
...   bar: str
...
>>> A().foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'A' object has no attribute 'foo'
```

If it's an issue with static checkers, then it can be guarded by `if TYPE_CHECKING`. Either way I'll add a comment that a potential `AttributeError` is expected and wanted. You can have your cake and eat it too, here 😄